bindings/matrix-sdk-crypto-ffi/src/lib.rs

```diff
@@ -350,6 +353,94 @@ impl<T: Fn(i32, i32)> ProgressListener for T {
     }
 }
 
+/// An encryption algorithm to be used to encrypt messages sent to a room.
+pub enum EventEncryptionAlgorithm {
+    /// Olm version 1 using Curve25519, AES-256, and SHA-256.
+    OlmV1Curve25519AesSha2,
+    /// Megolm version 1 using AES-256 and SHA-256.
+    MegolmV1AesSha2,
+}
+
+impl From<EventEncryptionAlgorithm> for RustEventEncryptionAlgorithm {
+    fn from(a: EventEncryptionAlgorithm) -> Self {
+        match a {
+            EventEncryptionAlgorithm::OlmV1Curve25519AesSha2 => {
+                RustEventEncryptionAlgorithm::OlmV1Curve25519AesSha2
+            }
+            EventEncryptionAlgorithm::MegolmV1AesSha2 => {
+                RustEventEncryptionAlgorithm::MegolmV1AesSha2
+            }
+        }
+    }
+}
+
+/// Who can see a room's history.
+pub enum HistoryVisibility {
+    /// Previous events are accessible to newly joined members from the point
+    /// they were invited onwards.
+    ///
+    /// Events stop being accessible when the member's state changes to
+    /// something other than *invite* or *join*.
+    Invited,
+
+    /// Previous events are accessible to newly joined members from the point
+    /// they joined the room onwards.
+    /// Events stop being accessible when the member's state changes to
+    /// something other than *join*.
+    Joined,
+
+    /// Previous events are always accessible to newly joined members.
+    ///
+    /// All events in the room are accessible, even those sent when the member
+    /// was not a part of the room.
+    Shared,
+
+    /// All events while this is the `HistoryVisibility` value may be shared by
+    /// any participating homeserver with anyone, regardless of whether they
+    /// have ever joined the room.
+    WorldReadable,
+}
+
+impl From<HistoryVisibility> for RustHistoryVisibility {
+    fn from(h: HistoryVisibility) -> Self {
+        match h {
+            HistoryVisibility::Invited => RustHistoryVisibility::Invited,
+            HistoryVisibility::Joined => RustHistoryVisibility::Joined,
+            HistoryVisibility::Shared => RustHistoryVisibility::Shared,
+            HistoryVisibility::WorldReadable => RustHistoryVisibility::Shared,
+        }
+    }
+}
+
+/// Settings for an encrypted room.
+///
+/// This determines the algorithm and rotation periods of a group session.
+pub struct EncryptionSettings {
```

Comment:
```suggestion
#[derive(uniffi::Object)]
pub struct EncryptionSettings {
```

doesn't do what you want?

Reply:
It does for the `EncryptionSettings`. But I would need to put the `share_room_key()` method, which takes this struct as an argument, under the impl block which uses `uniffi::export`. This doesn't work because `uniffi::export` doesn't support yet `Result` as a return value.

I tried to use the macro stuff, run into that blocker :smiling_face_with_tear:.